mitiq/calibration/calibration.py

```diff
@@ -0,0 +1,204 @@
+                )
+                expval_executor = bitstring_executor_to_expval_executor(
+                    self.executor, most_probable_bitstring
+                )
+                mitigated_expval = strategy.mitigation_function(
+                    circuit, expval_executor
+                )
+                mitigated[strategy.technique]["results"].append(
+                    {
+                        "circuit_type": circuit_data.type,
+                        "mitigated_value": mitigated_expval,
+                        **strategy.technique_params,
+                    }
+                )
+            expvals.append(
+                {
+                    "noisy_dist": noisy_distribution,
+                    "ideal_dist": ideal_distribution,
+                    "mitigated": mitigated,
+                }
+            )
+        return expvals
+
+    def compute_improvements(
+        self, experiment_results: list[dict[str, Any]]
+    ) -> None:
+        """Compute the improvement factors for each calibration circuit that
+        was run."""
+        for result in experiment_results:
+            ideal_dist = result["ideal_dist"]
+            ideal_expval = max(ideal_dist.values())
+            noisy_dist = result["noisy_dist"]
+            noisy_expval = max(noisy_dist.values())
+            for di in result["mitigated"].values():
+                results = di["results"]
+                mitigated_vals = list(
+                    map(lambda di: di["mitigated_value"], results)
+                )
+                method_improvement_factor = abs(
+                    noisy_expval - ideal_expval
+                ) / sqrt(
+                    len(mitigated_vals)
+                    * sum(
+                        (mitigated_val - ideal_expval) ** 2
+                        for mitigated_val in mitigated_vals
+                    )
+                )
+                di["method_improvement_factor"] = method_improvement_factor
+
+    def get_optimal_strategy(self, results: list[dict[str, Any]]) -> str:
+        """Finds the optimal error mitigation strategy using the improvement
+        factors calculated, and stored in `self.results`."""
+        best_val = 0.0
+        best_key = ""
+        for result in results:
+            ideal_dist = result["ideal_dist"]
+            ideal = max(ideal_dist.values())
+            noisy_dist = result["noisy_dist"]
+            noisy = max(noisy_dist.values())
+            unmitigated_error = abs((ideal - noisy) / ideal)
+            for method, di in result["mitigated"].items():
+                for res in di["results"]:
+                    mitigated_expval = res["mitigated_value"]
+                    diff = abs((ideal - mitigated_expval) / ideal)
+                    error_diff = abs(unmitigated_error - diff)
+                    if error_diff > best_val:
+                        best_val = error_diff
+                        best_key = method
```

Comment:
```suggestion
            for method, di in result["mitigated"].items():
               improvement_factor = di["improvement_factor]
               if error_diff > best_val:
                   best_val = improvement_factor
                   best_key = method
```
I forgot to add this to the above review.
This is just a sketch but the message is that `get_optimal_strategy()` should not re-compute the improvement factors.

Reply:
Ah yeah I see what you mean. Originally I was trying to get performance metrics for each result so we could know what parameters went along with the best performance. The suggestion you've made would be to find the best improvement factor for the EM method, but not any parameters.

Reply:
My main suggestion is to avoid re-computing the improvement factors since, in principle, we have a method `compute_improvements()` for doing it.  Actually it would be much better to know the optimal parameters of a technique. So my code suggestion is probably not good enough and we should improve it. 

The workflow that I have in mind is:

1.  `compute_improvements()`, computes the improvement factors for all combinations of circuit and strategy (strategy = technique + parameters). Schematically something like: **(CircuitData, Strategy) ==> Improvement Factor.** 

2. `get_optimal_strategy()` simply picks the `Strategy` with maximum improvement factor. In a future iteration of the code, the method `get_optimal_strategy() ` could do more advanced things based on the use-case circuit of the user. E.g., picking the best  `Strategy`  after filtering a subset of results associated to benchmark circuits with say $n$ qubits, or with depth $d$, etc...